Fix Bundler vendoring when a subdirectory is configured

deivid-rodriguez opened the pull request:
The actual fix is 51096421a9db36fcfcdf06131058374037681f16, but I also included some spec changes to be able to test it.

Previously we were passing an absolute path hardcoded to the root of the target repository as the cache path. Now we pass a relative path, to which we prepend the configured subdirectory if any.

Fixes #3736.

jurre reviewed (approved):
Thanks, that makes a lot of sense to me! Assuming CI passes, LGTM

deivid-rodriguez commented:
Yeah, CI definitely needed since I only run specs manually for Bundler, yet I changed helpers in common, so let's see :)

I also tested this on https://github.com/gewo/dependabot-bundler-vendor, and the dry-run scripts seem to behave fine now.

I didn't create actual PRs, just tried the dry-run script. But I can do that before merging just to make sure. The only difference to real PRs should be that from my machine, dependabot will add an extra platform to the lockfile, it seems.

## Before

<details>

<summary>$ bin/dry-run.rb bundler gewo/dependabot-bundler-vendor --clone ✅ </summary>

```
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
=> cloning into /home/dependabot/dependabot-core/tmp/gewo/dependabot-bundler-vendor
=> parsing dependency files
=> updating 1 dependencies: minitest

=== minitest (5.14.3)
 => checking for updates 1/1
🌍 https//rubygems.org:443/api/v1/versions/minitest.json
 => latest available version is 5.16.3
 => latest allowed version is 5.16.3
 => requirements to unlock: own
 => requirements update strategy: bump_versions
 => updating minitest from 5.14.3 to 5.16.3

    ± Gemfile.lock
    ~~~
    4c4
    <     minitest (5.14.3)
    ---
    >     minitest (5.16.3)
    6a7
    >   aarch64-linux
    ~~~
deleted vendor/cache/minitest-5.14.3.gem
added vendor/cache/minitest-5.16.3.gem
🌍 Total requests made: '1'
🎈 Package manager version log: {:ecosystem=>"bundler", :package_managers=>{"bundler"=>"2"}}
```
</details>

<details>
<summary>$ bin/dry-run.rb bundler gewo/dependabot-bundler-vendor --dir acceptance --clone ❌</summary>

```
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
=> cloning into /home/dependabot/dependabot-core/tmp/gewo/dependabot-bundler-vendor
=> parsing dependency files
=> updating 1 dependencies: minitest

=== minitest (5.14.3)
 => checking for updates 1/1
🌍 https//rubygems.org:443/api/v1/versions/minitest.json
 => latest available version is 5.16.3
 => latest allowed version is 5.16.3
 => requirements to unlock: own
 => requirements update strategy: bump_versions
 => updating minitest from 5.14.3 to 5.16.3

    ± Gemfile.lock
    ~~~
    4c4
    <     minitest (5.14.3)
    ---
    >     minitest (5.16.3)
    6a7
    >   aarch64-linux
    ~~~
🌍 Total requests made: '1'
🎈 Package manager version log: {:ecosystem=>"bundler", :package_managers=>{"bundler"=>"2"}}
```
</details>(note the missing "added/deleted" logging in the output)

## After

<details>
<summary>$ bin/dry-run.rb bundler gewo/dependabot-bundler-vendor --clone ✅ </summary>

```
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
=> cloning into /home/dependabot/dependabot-core/tmp/gewo/dependabot-bundler-vendor
=> parsing dependency files
=> updating 1 dependencies: minitest

=== minitest (5.14.3)
 => checking for updates 1/1
🌍 https//rubygems.org:443/api/v1/versions/minitest.json
 => latest available version is 5.16.3
 => latest allowed version is 5.16.3
 => requirements to unlock: own
 => requirements update strategy: bump_versions
 => updating minitest from 5.14.3 to 5.16.3

    ± Gemfile.lock
    ~~~
    4c4
    <     minitest (5.14.3)
    ---
    >     minitest (5.16.3)
    6a7
    >   aarch64-linux
    ~~~
deleted vendor/cache/minitest-5.14.3.gem
added vendor/cache/minitest-5.16.3.gem
🌍 Total requests made: '1'
🎈 Package manager version log: {:ecosystem=>"bundler", :package_managers=>{"bundler"=>"2"}}
```
</details>

<details>
<summary>$ bin/dry-run.rb bundler gewo/dependabot-bundler-vendor --dir /acceptance --clone ✅ </summary>

```
To use retry middleware with Faraday v2.0+, install `faraday-retry` gem
=> cloning into /home/dependabot/dependabot-core/tmp/gewo/dependabot-bundler-vendor
=> parsing dependency files
=> updating 1 dependencies: minitest

=== minitest (5.14.3)
 => checking for updates 1/1
🌍 https//rubygems.org:443/api/v1/versions/minitest.json
 => latest available version is 5.16.3
 => latest allowed version is 5.16.3
 => requirements to unlock: own
 => requirements update strategy: bump_versions
 => updating minitest from 5.14.3 to 5.16.3

    ± Gemfile.lock
    ~~~
    4c4
    <     minitest (5.14.3)
    ---
    >     minitest (5.16.3)
    6a7
    >   aarch64-linux
    ~~~
deleted vendor/cache/minitest-5.14.3.gem
added vendor/cache/minitest-5.16.3.gem
🌍 Total requests made: '1'
🎈 Package manager version log: {:ecosystem=>"bundler", :package_managers=>{"bundler"=>"2"}}
```
</details>
